### PR TITLE
add time_zones tables

### PR DESCRIPTION
**Pull request by bharatnc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add `time_zones` table.

Detailed description / Documentation draft:

This table adds a new table called the `time_zones` table:

```
:) select * from system.time_zones limit 10;

SELECT *
FROM system.time_zones
LIMIT 10

┌─time_zone───────────┐
│ America/Thunder_Bay │
│ Africa/Blantyre     │
│ Africa/Lusaka       │
│ America/Managua     │
│ Europe/Chisinau     │
│ Etc/GMT-7           │
│ Pacific/Midway      │
│ Asia/Irkutsk        │
│ America/Panama      │
│ Etc/GMT+5           │
└─────────────────────┘

10 rows in set. Elapsed: 0.012 sec.
```

The available timezones are parsed from under
`ClickHouse/contrib/cctz/testdata/zoneinfo` through a script
and the table content is constructed from the
`StorageSystemTimeZones.generated.cpp`.

I was able to test this locally and was able to get a list of time
zones as shown above.

This Addresses: #12444

### CODE REVIEW

File: src/Storages/System/StorageSystemTimeZones.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+NamesAndTypesList StorageSystemTimeZones::getNamesAndTypes()
+{
+    return {
+        {"time_zone", std::make_shared<DataTypeString>()},
+    };
+}
+
+void StorageSystemTimeZones::fillData(MutableColumns & res_columns, const Context &, const SelectQueryInfo &) const
+{
+    std::vector<const char *> time_zones;
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemTimeZones.cpp:20]: Maybe we can do it in single loop, without temporary array?

**Review by alexey-milovidov — COMMENTED** [review]
There is one slight deficiency in this implementation.

So, the list of timezones is generated at release time and committed to sources.
But when we build ClickHouse without a release, the list of timezones actually used may be different.
Instead, we can obtain the list of timezones during the build and generate source file in build directory.

Why it was done that way for system.contributors? Because we cannot obtain the list of contributors directly from source tree during build process. It requires git. But git is not necessary available during build - for example, if we build from source tarballs. So, for system.contributors it is better to pre-generate data.

But for system.time_zones, we don't have to pre-generate data. It's better to get the actual list of timezones that was used during build. See `contrib/cctz-cmake/CMakeLists.txt`

File: src/Storages/System/StorageSystemTimeZones.cpp (hunk 1 above)
Comment (bharatnc) [reply src/Storages/System/StorageSystemTimeZones.cpp]: yes definitely, should be fixed now.

**Comment by bharatnc** [comment]
> There is one slight deficiency in this implementation.
> 
> So, the list of timezones is generated at release time and committed to sources.
> But when we build ClickHouse without a release, the list of timezones actually used may be different.
> Instead, we can obtain the list of timezones during the build and generate source file in build directory.
> 
> Why it was done that way for system.contributors? Because we cannot obtain the list of contributors directly from source tree during build process. It requires git. But git is not necessary available during build - for example, if we build from source tarballs. So, for system.contributors it is better to pre-generate data.
> 
> But for system.time_zones, we don't have to pre-generate data. It's better to get the actual list of timezones that was used during build. See `contrib/cctz-cmake/CMakeLists.txt`

1. I have refactored the PR to directly generate the `StorageSystemTimeZones.generated.cpp` file from within the `cctz-cmake/CMakeList.txt` file.

2. For a OS_LINUX we are now using all the time-zones that's the same as the ones that's being baked into the ClickHouse binary.

3. Regarding 2 - this also means that this list would not work for other systems like OS_X, BSD etc. I'm checking for the presence of the generated`StorageSystemTimeZones.generated.cpp` file before attaching the `system.time_zones` table which means, this table won't exist currently for other OSes except OS_LINUX. Do you recommend a work around here or this sounds good to you?  Thank you for your feedback.

**Comment by alexey-milovidov** [comment]
> Do you recommend a work around here or this sounds good to you?

I think it's Ok.

**Comment by bharatnc** [comment]
> Looks like you forgot to commit test reference file.

oops, yes, just added.

**Comment by filimonov** [comment]
Just a reminder / related - i think that can be merged https://github.com/ClickHouse/ClickHouse/pull/13648

**Comment by bharatnc** [comment]
> > Looks like you forgot to commit test reference file.
> 
> oops, yes, just added.

Test fails with:
```bash
2020-08-20 22:32:54 
2020-08-20 22:32:54 Received exception from server (version 20.8.1):
2020-08-20 22:32:54 Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table system.time_zones doesn't exist.. 
2020-08-20 22:32:54 
2020-08-20 22:32:54 , result:
2020-08-20 22:32:54 
2020-08-20 22:32:54 
```

Will figure out what's going on - not sure if the table get's attached in the way I expect it to.

**Comment by bharatnc** [comment]
The build uses the binary: [clickhouse-server_20.8.1.4436_all.deb](https://clickhouse-builds.s3.yandex.net/13880/1cb9fe21b911c6bccfaddad1a00306927da9a8ac/clickhouse_build_check/gcc-9_relwithdebuginfo_none_bundled_unsplitted_disable_False_deb/clickhouse-server_20.8.1.4436_all.deb) which I downloaded and installed locally on my linux dev machine and I'm able to run that failing query just fine:

```bash
:) SELECT if ((SELECT count(*) FROM system.time_zones) = 562, 'ok', 'fail');

SELECT if(
    (
        SELECT count(*)
        FROM system.time_zones
    ) = 562, 'ok', 'fail')

┌─if(equals(_subquery1, 562), 'ok', 'fail')─┐
│ ok                                        │
└───────────────────────────────────────────┘
```

Not entirely sure why it's failing.

**Comment by alexey-milovidov** [comment]
Stress test (undefined) - fixed in #13980
Performance - sporadic changes, Ok.

**Comment by alexey-milovidov** [comment]
There a few minor changes required... I will add them in another PR and this PR will be merged automatically.

File: tests/queries/0_stateless/01455_time_zones.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-- There are currently 562 timezones which are used from the binary embedded inside the ClickHouse binary
+-- Refer: contrib/cctz-cmake/CMakeLists.txt for the complete list. The count should match when queried.
+-- SELECT count(*)
+-- FROM system.time_zones
+--
+-- ┌─count()─┐
+-- │     562 │
+-- └─────────┘
+SELECT if ((SELECT count(*) FROM system.time_zones) = 562, 'ok', 'fail');
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01455_time_zones.sql:9]: In recent commit we started to generate the complete list on the fly. And the number of timezones may change. I will just check that there are more than 500 :)

File: src/Storages/ya.make [hunk 3]
```diff
@@ -182,6 +182,8 @@ SRCS(
     System/StorageSystemTableEngines.cpp
     System/StorageSystemTableFunctions.cpp
     System/StorageSystemTables.cpp
+    System/StorageSystemTimeZones.cpp
+    System/StorageSystemTimeZones.generated.cpp
```
Comment (alexey-milovidov) [inline src/Storages/ya.make:186]: Actually ya.make is used for "alien build system" that are using timezones in completely different way... I will exclude system.timezones for that build.

File: src/Storages/System/StorageSystemTimeZones.generated.cpp [hunk 4]
```diff
@@ -0,0 +1,565 @@
+// autogenerated by ClickHouse/contrib/cctz-cmake/CMakeLists.txt
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemTimeZones.generated.cpp:1]: This should not be committed.

File: contrib/cctz-cmake/CMakeLists.txt [hunk 5]
```diff
@@ -24,8 +24,14 @@ if (USE_INTERNAL_CCTZ)
         target_compile_definitions (cctz PRIVATE __USE_BSD linux _XOPEN_SOURCE=600)
     endif ()
 
+    # Related to time_zones table:
+    # StorageSystemTimeZones.generated.cpp is autogenerated each time during a build
+    # data in this file will be used to populate the system.time_zones table, this is specific to OS_LINUX
+    # as the library that's built using embedded tzdata is also specific to OS_LINUX
+    set(SYSTEM_STORAGE_TZ_FILE "../../src/Storages/System/StorageSystemTimeZones.generated.cpp")
```
Comment (alexey-milovidov) [inline contrib/cctz-cmake/CMakeLists.txt:31]: I will try to use ${CMAKE_CURRENT_BINARY_DIR} just in case...

**Review by alexey-milovidov — APPROVED** [review]

File: src/Storages/System/StorageSystemTimeZones.generated.cpp (hunk 4 above)
Comment (bharatnc) [reply src/Storages/System/StorageSystemTimeZones.generated.cpp:1]: I see -  interesting, I realized that the tests fail with `system.time_zones` table not found without committing this file. Besides files like the [StorageSystemContributors.generated.cpp](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/System/StorageSystemContributors.generated.cpp), [StorageSystemBuildOptions.generated.cpp.in](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/System/StorageSystemBuildOptions.generated.cpp.in) are being committed though. 

Though I realized that downloading the exact binary being build and running locally actually shows the systems.time_zones table all fine.

**Comment by alexey-milovidov** [comment]
Merged in #14030.

**Pull request closed without merging**